feat(rxCheckbox): add rxCheckbox component

CITguy opened the pull request:
Per necessity of #884, this directive can stand alone for use elsewhere in the framework.

parlarjb commented:
Could you include some screenshots?

CITguy commented:
![screen shot 2015-05-06 at 9 56 33 am](https://cloud.githubusercontent.com/assets/545605/7495863/3089deac-f3d6-11e4-8af5-b9003fec1ac3.png)

CITguy commented:
Rebased

megalord commented:
I'm curious, why are the attributes being documented in the demo markup as opposed to via ngdocs?  That seems to be the standard from what I've seen.

megalord commented on src/rxCheckbox/rxCheckbox.less:
```diff
@@ -0,0 +1,99 @@
+ */
+.rxCheckbox {
+  display: inline-block;
+  width: @rxCheckbox-size; // OK
```
Can you remove all of the `// OK` comments?

CITguy replied:
Ah, good call.

megalord commented on src/rxCheckbox/rxCheckbox.page.js:
```diff
@@ -0,0 +1,103 @@
+     * @function
+     * @description make sure checkbox is deselected/unchecked
+     */
+    deselect: {
```
The `rxForm` version of a checkbox uses "unselect" instead of "deselect".  While I'm not partial to either, it may be confusing for anyone migrating to have to change their tests.  It would be easier if all they had to do is change
```javascript
rxForm.checkbox.initialize(...)
```
to
```javascript
rxCheckbox.initialize(...)
```

CITguy replied:
Keep in mind, that this PR is to support upcoming changes to rxForm.

CITguy replied:
Also, http://english.stackexchange.com/questions/18465/unselect-or-deselect

CITguy replied:
Should rxForm use rxCheckbox, then `rxForm.checkbox.initialize(...)` will initialize an rxCheckbox item.

CITguy replied:
Sorry, I misunderstood.  With the upcoming changes to rxForm, there's going to be a considerable update. `rxForm.checkbox.initialize(...)` only applies to old rxForm functionality (using native checkboxes).  As developers transition to using the new rxForm functionality (using newly styled checkboxes), they would naturally have to transition to the new page objects as well.

megalord replied:
Yeah, there will be some big changes to rxForm, and my concern is that something like this is not obvious and will be lost amidst the other changes, so the tests will break in unexpected places.  Perhaps as the larger change lands, it would be wise to have a migration guide where the big picture changes as well as details like this can be documented.

CITguy replied:
I think we can revisit this topic in the rxForm updates, if needed.

Droogans replied:
Since I'm the one responsible for this odd naming convention, I feel like I should at least explain why I went with unselect vs. deselect. At first, there was `select` for dropdowns and radio inputs, and `check` for checkboxes.

I didn't like the variation in language. It made it harder for me to remember what to call. So I switched everything over to `select`, and in the process, saw `uncheck` and thought...."should this be `unselect`? What about `deselect`?". After thinking about it, I went with something that worked everywhere (since there was no `decheck`), ending up settling on `unselect`, which may need to change.

TL;DR - I'm all for switching everything over to `deselect`, since there's no more variation between verbs.

Droogans replied:
Also, @megalord, @CITguy and I discussed what'll likely happen to the page objects once all the rxForm components get broken out of it. I think we'll be using `rxForm` as a "require everything please" page object, so it'll keep backwards compatibility. It'll also make pulling in every form input type less lines of code (one `require` instead of three or four).

CITguy commented:
@megalord, I need to apply a styling to the first cell in the row so that the value doesn't wrap.  I cannot do this via markdown.  I addition, I wanted to keep non-demo markup to bhe bottom so that when a consumer looks at the "Markup" tab, they see the demo code first.

megalord commented on src/rxCheckbox/rxCheckbox.spec.js:
```diff
@@ -0,0 +1,21 @@
+        el = helpers.createDirective(validTemplate, compile, scope);
+    });
+
+    /* reserved for potential future specs */
```
I'm not a big fan of having a spec file without any tests in it.  I say either add tests for the watch on `ng-disabled` or remove the file entirely.

CITguy replied:
I only kept it in case someone pointed out something that required testing in this review. If there's nothing worth testing, then I can remove it. I'll defer to @Droogans on this one.

megalord commented:
@CITguy ngdocs doesn't use markdown - it's generated from the comments that are placed before each `.directive`, `.service`, etc.

CITguy commented:
Added "DO NOT MERGE" because functionality is needed to remove the wrapper markup on a destroy (thanks for pointing this out @megalord). This is needed when an `ng-if` or similar removes the `<input rx-checkbox ... />` element from the DOM.

CITguy commented:
* Rebased
* removed `rxCheckbox.spec.js` (no functionality to test)
* added `$destroy` support
* updated midways and page object
* updated demo and documentation

CITguy commented:
![rxcheckbox-local-20150508](https://cloud.githubusercontent.com/assets/545605/7540599/97ca2ca8-f574-11e4-9908-1f2ac73949c4.png)

CITguy commented:
*unchecked (invalid)*
![screen shot 2015-05-08 at 11 25 46 am](https://cloud.githubusercontent.com/assets/545605/7540673/0d89df7e-f575-11e4-8ac8-ae7391e7aac9.png)

*checked (valid), shows next checkbox via `ng-show` logic*
![screen shot 2015-05-08 at 11 25 54 am](https://cloud.githubusercontent.com/assets/545605/7540672/0d86b98e-f575-11e4-9e23-69a04862d82e.png)

*both checked and valid*
![screen shot 2015-05-08 at 11 26 03 am](https://cloud.githubusercontent.com/assets/545605/7540671/0d84dc40-f575-11e4-88be-8d92174b3839.png)

CITguy commented:
**`$destroy` support**
*unchecked, second checkbox is present*
![screen shot 2015-05-08 at 11 26 11 am](https://cloud.githubusercontent.com/assets/545605/7540700/4f6925bc-f575-11e4-9cdd-5681a2944efd.png)

*checked, second checkbox (and stylistic markup) removed from DOM*
![screen shot 2015-05-08 at 11 26 17 am](https://cloud.githubusercontent.com/assets/545605/7540701/4f6a32c2-f575-11e4-96ff-4b1ad2aa1a80.png)

parlarjb commented on src/rxCheckbox/README.md:
````diff
@@ -0,0 +1,28 @@
+```html
+<span ng-show="isShown">
+  <input rx-checkbox
+         type="checkbox"
````
Could the `rx-checkbox` directive automatically set the `type="checkbox"`, in the `compile` step?

CITguy replied:
Unfortunately, I cannot use both `compile` and `link` in the DDO. Considering that I need to perform DOM manipulation, I need to keep `link`. I tried to automatically set the `type="checkbox"` in the link function, but for some reason, it breaks binding with the scope.

CITguy replied:
In addition, if for some reason that the directive doesn't bind properly to the DOM, the `type="checkbox"` will act as a fallback for graceful degradation.

megalord replied:
Just curious, are you opposed to returning the `link` function from `compile`?

CITguy replied:
I had a chat with @parlarjb about it. He convinced me to convert to using `compile` and returning the function. I'll be submitting an update to the PR soon.

parlarjb commented on src/rxCheckbox/README.md:
````diff
@@ -0,0 +1,28 @@
+```
+
+It is highly recommended that you use `ng-show` and `ng-hide` for purposes of display logic. Because of the way that `ng-if` and `ng-switch`
+directives behave with scope, they may introduce unnecessary complexity in your code.
````
:+1:

parlarjb commented on src/rxCheckbox/rxCheckbox.js:
```diff
@@ -0,0 +1,34 @@
+    return {
+        restrict: 'A',
+        scope: {
+            ngDisabled: '=?'
```
This doesn't cause problems with the existing `ngDisabled` directive that ships with Angular?

CITguy replied:
Nope, just watches the value so that a `scope.$watch()` can add/remove a `disabled` attribute on the wrapper element.

parlarjb commented on src/rxCheckbox/rxCheckbox.js:
```diff
@@ -0,0 +1,34 @@
+
+            element.wrap(wrapper);
+            element.after(fakeCheckbox);
+
```
Rather that doing `element.parent()` on every $watch call below, could you do `wrapper = angular.element(wrapper)`,
and then inside the `$watch` you an do `wrapper.attr('disabled', true);` and `wrapper.removeAttr('disabled')`.
Then in the `$destroy` you can also do `wrapper[0].remove();`

CITguy replied:
ah, I will do that. much cleaner.

CITguy commented:
Updated per comments:

* using `compile` rather than `link` so that the directive automatically sets `type="checkbox"`
* applying logic to wrapper element rather than `element.parent()` repetition

Additionally, I reorganized some of the documentation to make it easier to consume and navigate. The examples still work same as seen in previous screenshots.

![rxcheckbox-local-20150511-1](https://cloud.githubusercontent.com/assets/545605/7574270/ab9f58aa-f7f1-11e4-81d5-8f69fc19235a.png)

parlarjb commented:
Could you put back in `.spec.js` to ensure that `type="checkbox"` is being correctly added, as well as checking that your custom `ng-disabled` logic is being applied?

CITguy commented:
@parlarjb Midways have been updated to test `ng-disabled` and `type="checkbox"` functionality. I didn't feel like this warranted the need for a spec file as the directive isn't adding any functionality on `scope` or the controller. It merely manipulates the DOM based on data binding. @Droogans do you have any input on testing via unit tests versus midways?

parlarjb commented:
LGTM

CITguy commented:
Based on feedback for `rxRadio`, I need to add logic to support the native `disabled` attribute in addition to `ng-disabled`.

CITguy commented:
* added support for native `disabled` attribute

*Checkbox States - updated to include native `disabled` examples*
![screen shot 2015-05-13 at 11 29 57 am](https://cloud.githubusercontent.com/assets/545605/7615503/ac9a6a24-f963-11e4-9152-1a4015a7b471.png)

megalord commented on src/rxCheckbox/rxCheckbox.js:
```diff
@@ -0,0 +1,45 @@
+                if (element.attr('disabled')) {
+                    wrapper.addClass(disabledClass);
+                }
+                if (scope.ngDisabled) {
```
Should this be `attr.ngDisabled`? If the expression passed to `ng-disabled` is initially false, the watch should still be applied.

CITguy replied:
I see what you're saying. I was trying to check for the presence of the variable, not its value. I'll get a quick update submitted.

CITguy commented on src/rxCheckbox/rxCheckbox.js:
```diff
@@ -0,0 +1,45 @@
+                if (attrs.disabled) {
+                    wrapper.addClass(disabledClass);
+                }
+                if (angular.isDefined(scope.ngDisabled)) {
```
@megalord, how's this look?

megalord replied:
Well, this doesn't work for a value of `undefined`, which would work on a normal checkbox input.  At this point though, I feel I'm getting too nit-picky, so feel free to keep it as is unless someone else chimes in.

parlarjb replied:
`if (_.has(attrs, 'ngDisabled')) {` would probably be the most correct way to do it

CITguy commented:
Updated to use `if (_.has(attrs, 'ngDisabled'))`

megalord commented:
LGTM
